# Week 4 Lab — Inheritance & Polymorphism (OOP III)

**192-201 Advanced Computer Programming with Generative AI**

Reference: *Think Python, 3rd Edition* — Chapter 17 (Inheritance) + Chapter 16 (Polymorphism) + lecturer supplement on `super()`.
Builds on **Week 2** (classes) and **Week 3** (encapsulation).

---

### About this lab

This is a **guided practice notebook** — it is **not graded**. It follows the **five ideas** from today's slides. Each part is tagged with the slide concept it practises, so you can look back if you get stuck:

> **① Inheritance · ② is-a · ③ Override + super() · ④ Polymorphism · ⑤ is-a vs has-a**

Work through the parts **in order**. For each part: read the guidance, **run every code cell**, then complete each **Your turn** / **TODO** task.

> **Run a cell:** click it, then press **Shift + Enter**.

> **AI-use policy — Level 1–2 this week:** AI may *explain* ideas and *suggest* structure, but **you must be able to explain every line you keep**. Do not paste code you don't understand.

## Part 0 — Recap: one class on its own

Here is a single `Shape` class, like the ones from Weeks 2–3. Run it — then notice the problem: if we also want a `Circle` and a `Square`, we'd copy most of this code. Inheritance fixes that.

In [29]:
class Shape:
    def __init__(self, name):
        self.name = name

    def describe(self):
        print(f'I am a {self.name}.')

s = Shape('shape')
s.describe()          # I am a shape.

I am a shape.


## Part 1 — Inheritance: build a child class  ·  ① Inheritance

> **From the slides — Concept 1 (Inheritance):** *define a new class as a modified version of an existing one.* The trick is putting the parent's name **in parentheses**.

`Circle` inherits everything from `Shape` — including `describe()` — without rewriting it. Run this:

In [30]:
class Circle(Shape):        # Circle inherits from Shape
    pass                    # adds nothing yet — but already has describe()

c = Circle('circle')
c.describe()                # I am a circle.   <- inherited method!

I am a circle.


> Notice `Circle` has a working `describe()` method that we **never wrote** — it came from `Shape`. That is inheritance.

**Your turn.** Make a class `Square` that inherits from `Shape` (just like `Circle`). Create a `Square('square')` and call `describe()` on it.

In [31]:
class Square(Shape):
    pass

s = Square('square')
s.describe()

I am a square.


## Part 2 — The “is-a” test  ·  ② is-a

> **From the slides — Concept 2 (is-a):** *inherit only when the child truly **is a** kind of the parent.*

You don't run code for this part — you **think**. For each pair below, decide: is it **is-a** (use inheritance) or **has-a** (don't)? Double-click this cell and write your answer after each arrow.

1. A `Dog` and an `Animal` → *is-a*
2. A `Car` and an `Engine` → *has-a*
3. A `Circle` and a `Shape` → *is-a*
4. A `Playlist` and a `Song` → *has-a*
5. A `Manager` and an `Employee` → *is-a*

*(Answers: 1 is-a · 2 has-a · 3 is-a · 4 has-a · 5 is-a — check after you try.)*

## Part 3 — Give the child its own data with super()  ·  ③ super()

> **From the slides — Concept 3 (super()):** *let the child do everything the parent does first, then add its own part.*

A circle needs a **radius** as well as a name. In `__init__`, we call `super().__init__(...)` to run the parent's setup (which stores the name), then add the radius. Run this:

In [32]:
class Circle(Shape):
    def __init__(self, radius):
        super().__init__('circle')   # run Shape.__init__ -> sets self.name
        self.radius = radius         # then add the child's own attribute

c = Circle(5)
c.describe()          # I am a circle.   (from the parent)
print(c.radius)       # 5               (from the child)

I am a circle.
5


> Without `super().__init__('circle')`, the object would have **no** `name`, and `describe()` would fail. `super()` reuses the parent's work instead of copying it.

**Your turn.** Write a `Rectangle(Shape)` whose `__init__` takes `width` and `height`. Call `super().__init__('rectangle')`, then store `self.width` and `self.height`. Create one and call `describe()`.

In [33]:
class Rectangle(Shape):
    def __init__(self, width, height):
        super().__init__('rectangle')
        self.width = width
        self.height = height

r = Rectangle(3, 4)
r.describe()

I am a rectangle.


## Part 4 — Overriding: change a method in the child  ·  ③ Override

> **From the slides — Concept 3 (Overriding):** *a child can define a method with the same name to replace the parent's version.*

We give `Shape` an `area()` method that returns `0`, then **override** it in `Circle` with the real formula. Run this:

In [21]:
class Shape:
    def __init__(self, name):
        self.name = name
    def area(self):
        return 0                     # generic default

class Circle(Shape):
    def __init__(self, radius):
        super().__init__('circle')
        self.radius = radius
    def area(self):                  # OVERRIDE: same name, new behaviour
        return 3.14159 * self.radius ** 2

print(Shape('x').area())    # 0        (parent's version)
print(Circle(5).area())     # 78.53975 (child's override)

0
78.53975


> Same method name, `area()`, but a `Circle` object runs the **child's** version. That is overriding.

**Your turn.** Add `Rectangle(Shape)` again, this time with an `area()` that returns `self.width * self.height`. Test it: a `Rectangle(3, 4)` should give `12`.

In [34]:
class Rectangle(Shape):
    def __init__(self, width, height):
        super().__init__('rectangle')
        self.width = width
        self.height = height

    def area(self):
        return self.width * self.height

r = Rectangle(3, 4)
print(r.area())

12


## Part 5 — Polymorphism: one loop, many types  ·  ④ Polymorphism

> **From the slides — Concept 4 (Polymorphism):** *different types share a method name, so one loop works on all of them.*

This is the payoff. Because **every** shape has an `area()` method, we can put different shapes in one list and loop over them — each object runs its own `area()`. First, here is the full hierarchy (the same one from the slides):

In [38]:
class Shape:
    def __init__(self, name):
        self.name = name
    def area(self):
        return 0

class Circle(Shape):
    def __init__(self, radius):
        super().__init__('circle')
        self.radius = radius
    def area(self):
        return 3.14159 * self.radius ** 2

class Rectangle(Shape):
    def __init__(self, width, height):
        super().__init__('rectangle')
        self.width = width
        self.height = height
    def area(self):
        return self.width * self.height

Now the polymorphic loop — **one line** handles every shape type:

In [39]:
shapes = [Circle(2), Rectangle(3, 4), Circle(1)]

for shape in shapes:
    print(f'{shape.name}: area = {shape.area():.2f}')

# circle: area = 12.57
# rectangle: area = 12.00
# circle: area = 3.14

circle: area = 12.57
rectangle: area = 12.00
circle: area = 3.14


> The loop never checks *which* shape it has. It just trusts every shape has `area()`. Add a new shape type tomorrow and this loop works unchanged — **that** is why polymorphism matters.

**Your turn.** Add a `Triangle(Shape)` class (`__init__` takes `base` and `height`; `area()` returns `0.5 * base * height`). Then add a triangle to the `shapes` list and re-run the loop — it should print the triangle's area with no other changes.

In [40]:
class Triangle(Shape):
    def __init__(self, base, height):
        super().__init__('triangle')
        self.base = base
        self.height = height

    def area(self):
        return 0.5 * self.base * self.height

shapes = [Circle(2), Rectangle(3, 4), Circle(1), Triangle(4, 5)]

for shape in shapes:
    print(f'{shape.name}: area = {shape.area():.2f}')

circle: area = 12.57
rectangle: area = 12.00
circle: area = 3.14
triangle: area = 10.00


## Part 6 — Guided build: sum the areas  ·  ④ Polymorphism

Put it all together. Write a function `total_area(shapes)` that returns the **sum** of the areas of a list of shapes. It should work for **any** mix of shape types — that is polymorphism doing the heavy lifting. Fill in the `# TODO`.

In [41]:
def total_area(shapes):
    total = 0
    for shape in shapes:
        total += shape.area()
    return total

my_shapes = [Circle(2), Rectangle(3, 4)]
print(round(total_area(my_shapes), 2))

24.57


**Check your work.** With `[Circle(2), Rectangle(3, 4)]` the total is `12.57 + 12.00 = 24.57`. Add another shape and confirm the total updates — you never have to change `total_area` itself.

## Part 7 — Design review (work in pairs)  ·  ⑤ is-a vs has-a

> **From the slides — Concept 5 (is-a vs has-a):** the real skill is choosing inheritance only when it fits.

With a partner, invent **one** new class hierarchy of your own (not shapes). Fill in the table. Then find **one** “has-a” relationship in your idea that should **not** use inheritance.

| Parent class | Two child classes | A method the children override | One “has-a” (do NOT inherit) |
|---|---|---|---|
| *(e.g. Animal)* | *(Dog, Cat)* | *(speak)* | *(a Dog has-a Collar)* |
| *(Vehical)* | *(Car, Motorcycle)* | *(move)* | *(Car has-an Engine)* |

## Part 8 — Learning with an AI assistant (Level 1–2)

This week AI may **explain** and **suggest structure**, but you must understand and be able to defend every line. Good questions:

- "Explain what `super().__init__()` does and why the child needs it."
- "What is the difference between overriding a method and overloading?"
- "Explain how one `for` loop can call `area()` on different shape types."
- "Suggest a sensible parent class for `Dog`, `Cat`, and `Bird` — and say why."

**Still not allowed:** pasting a whole solution you can't explain. If you keep an AI suggestion, be ready to explain it in the Week 8 code-defence.

## Before you finish — quick checklist

You should now be able to — and be able to **name** each idea:

- [ ] **① Inheritance** — make a child class with `class Child(Parent):`
- [ ] **② is-a** — decide when inheritance is the right choice
- [ ] **③ super()** — call the parent's `__init__` from the child
- [ ] **③ Override** — replace a parent method in the child
- [ ] **④ Polymorphism** — write one loop that treats many types the same
- [ ] **⑤ is-a vs has-a** — avoid inheriting when it's really “has-a”

**Assignment 1 is due this week** — your OOP submission (Weeks 2–4) should now use a small class hierarchy. Push it to your GitHub repo.

**Before Week 5:** next week is **Design & Refactoring with AI** — improving working code and building objects out of other objects (composition — the “has-a” you spotted today).

*Great work*